### ISSUE TITLE
Invalid SFTP packet size

### ISSUE THREAD

**Issue by stresbiz** [phpseclib/phpseclib#1606]
Hi there,

We are using phpseclib 1.0.19 and are seeing an error: Invalid SFTP packet size /phpseclib/Net/SFTP.php 3152

It looks like here is the code that is throwing the error in SFTP.php:

```
// 256 * 1024 is what SFTP_MAX_MSG_LENGTH is set to in OpenSSH's sftp-common.h
        if ($tempLength > 256 * 1024) {
            user_error('Invalid SFTP packet size');
            return false;
        }
```

**Comment by terrafrost** [comment]
What are you doing that's triggering the error?

Can you post the SSH2 logs? You can get them by doing `define('NET_SSH2_LOGGING', 2)` and then by doing `echo $sftp->getLog()` after the fact.

The SFTP logs might help too. Those can be gotten by doing `define('NET_SFTP_LOGGING', 2)` and by doing `$sftp->getSFTPLog()` after the fact.

I don't want _just_ the SFTP logs because the SSH2 logs could contain relevant info. The SFTP logs make it waaaay easier to parse SFTP info so they'd be nice too to save me some time!

**Comment by stresbiz** [comment]
We are downloading 100s of files using $sftp->get() and running it in a loop to download each file one by one from a remote directory.

It appears that we're running out of memory from the logs.  I upped the memory to 10GB, which prolonged the script and downloaded many more files, but it still failed before completing the batch: ini_set('memory_limit', '10240M');

Any suggestion on how to avoid running out of memory with phpseclib 1.0.19?

**Comment by terrafrost** [comment]
Just get me the SSH2 logs then. The SSH2 logs cap the size of the log file at 1MB - I thought the SFTP logs did as well but I guess they don't!

Thanks!

**Comment by stresbiz** [comment]
My SSH2 logs turned out to be 105MB, but perhaps it's because I keep appending the log to the file.  Here is where it starts failing at the end of the log:

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0004, network: 0s)
00000000  00:00:01:00:00:00:00:20:00:00:00:1c:65:00:00:00  ....... ....e...
00000010  0b:00:00:00:01:00:00:00:0b:45:6e:64:20:6f:66:20  .........End of 
00000020  66:69:6c:65:00:00:00:00                          file....

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0001, network: 0s)
00000000  00:00:01:00:00:00:00:20:00:00:00:1c:65:00:00:00  ....... ....e...
00000010  0c:00:00:00:01:00:00:00:0b:45:6e:64:20:6f:66:20  .........End of 
00000020  66:69:6c:65:00:00:00:00                          file....

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0001, network: 0s)
00000000  00:00:01:00:00:00:00:20:00:00:00:1c:65:00:00:00  ....... ....e...
00000010  0d:00:00:00:01:00:00:00:0b:45:6e:64:20:6f:66:20  .........End of 
00000020  66:69:6c:65:00:00:00:00                          file....

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.05, network: 0s)
00000000  00:00:01:00:00:00:00:20:00:00:00:1c:65:00:00:00  ....... ....e...
00000010  0e:00:00:00:01:00:00:00:0b:45:6e:64:20:6f:66:20  .........End of 
00000020  66:69:6c:65:00:00:00:00                          file....

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0001, network: 0s)
00000000  00:00:01:00:00:00:00:20:00:00:00:1c:65:00:00:00  ....... ....e...
00000010  0f:00:00:00:01:00:00:00:0b:45:6e:64:20:6f:66:20  .........End of 
00000020  66:69:6c:65:00:00:00:00                          file....

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0001, network: 0s)
00000000  00:00:01:00:00:00:00:20:00:00:00:1c:65:00:00:00  ....... ....e...
00000010  10:00:00:00:01:00:00:00:0b:45:6e:64:20:6f:66:20  .........End of 
00000020  66:69:6c:65:00:00:00:00                          file....

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0001, network: 0s)
00000000  00:00:01:00:00:00:00:20:00:00:00:1c:65:00:00:00  ....... ....e...
00000010  11:00:00:00:01:00:00:00:0b:45:6e:64:20:6f:66:20  .........End of 
00000020  66:69:6c:65:00:00:00:00                          file....

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0, network: 0s)
00000000  00:00:01:00:00:00:00:20:00:00:00:1c:65:00:00:00  ....... ....e...
00000010  12:00:00:00:01:00:00:00:0b:45:6e:64:20:6f:66:20  .........End of 
00000020  66:69:6c:65:00:00:00:00                          file....

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0, network: 0s)
00000000  00:00:01:00:00:00:00:20:00:00:00:1c:65:00:00:00  ....... ....e...
00000010  13:00:00:00:01:00:00:00:0b:45:6e:64:20:6f:66:20  .........End of 
00000020  66:69:6c:65:00:00:00:00                          file....

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0, network: 0s)
00000000  00:00:01:00:00:00:00:20:00:00:00:1c:65:00:00:00  ....... ....e...
00000010  14:00:00:00:01:00:00:00:0b:45:6e:64:20:6f:66:20  .........End of 
00000020  66:69:6c:65:00:00:00:00                          file....

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0001, network: 0s)
00000000  00:00:01:00:00:00:00:20:00:00:00:1c:65:00:00:00  ....... ....e...
00000010  15:00:00:00:01:00:00:00:0b:45:6e:64:20:6f:66:20  .........End of 
00000020  66:69:6c:65:00:00:00:00                          file....

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0001, network: 0s)
00000000  00:00:01:00:00:00:00:20:00:00:00:1c:65:00:00:00  ....... ....e...
00000010  16:00:00:00:01:00:00:00:0b:45:6e:64:20:6f:66:20  .........End of 
00000020  66:69:6c:65:00:00:00:00                          file....

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0001, network: 0s)
00000000  00:00:01:00:00:00:00:20:00:00:00:1c:65:00:00:00  ....... ....e...
00000010  17:00:00:00:01:00:00:00:0b:45:6e:64:20:6f:66:20  .........End of 
00000020  66:69:6c:65:00:00:00:00                          file....

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0001, network: 0s)
00000000  00:00:01:00:00:00:00:20:00:00:00:1c:65:00:00:00  ....... ....e...
00000010  18:00:00:00:01:00:00:00:0b:45:6e:64:20:6f:66:20  .........End of 
00000020  66:69:6c:65:00:00:00:00                          file....

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0001, network: 0s)
00000000  00:00:01:00:00:00:00:20:00:00:00:1c:65:00:00:00  ....... ....e...
00000010  19:00:00:00:01:00:00:00:0b:45:6e:64:20:6f:66:20  .........End of 
00000020  66:69:6c:65:00:00:00:00                          file....

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0001, network: 0s)
00000000  00:00:01:00:00:00:00:20:00:00:00:1c:65:00:00:00  ....... ....e...
00000010  1a:00:00:00:01:00:00:00:0b:45:6e:64:20:6f:66:20  .........End of 
00000020  66:69:6c:65:00:00:00:00                          file....

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0001, network: 0s)
00000000  00:00:01:00:00:00:00:20:00:00:00:1c:65:00:00:00  ....... ....e...
00000010  1b:00:00:00:01:00:00:00:0b:45:6e:64:20:6f:66:20  .........End of 
00000020  66:69:6c:65:00:00:00:00                          file....

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0001, network: 0s)
00000000  00:00:01:00:00:00:00:20:00:00:00:1c:65:00:00:00  ....... ....e...
00000010  1c:00:00:00:01:00:00:00:0b:45:6e:64:20:6f:66:20  .........End of 
00000020  66:69:6c:65:00:00:00:00                          file....

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0001, network: 0s)
00000000  00:00:01:00:00:00:00:20:00:00:00:1c:65:00:00:00  ....... ....e...
00000010  1d:00:00:00:01:00:00:00:0b:45:6e:64:20:6f:66:20  .........End of 
00000020  66:69:6c:65:00:00:00:00                          file....

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0001, network: 0s)
00000000  00:00:01:00:00:00:00:20:00:00:00:1c:65:00:00:00  ....... ....e...
00000010  1e:00:00:00:01:00:00:00:0b:45:6e:64:20:6f:66:20  .........End of 
00000020  66:69:6c:65:00:00:00:00                          file....

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.07, network: 0.0001s)
00000000  00:00:01:00:00:00:00:20:00:00:00:1c:65:00:00:00  ....... ....e...
00000010  1f:00:00:00:01:00:00:00:0b:45:6e:64:20:6f:66:20  .........End of 
00000020  66:69:6c:65:00:00:00:00                          file....

-> NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0003, network: 0.0001s)
00000000  00:00:00:00:00:00:00:2d:00:00:00:29:04:00:00:00  .......-...)....
00000010  01:00:00:00:20:38:31:63:66:36:39:63:37:63:30:39  .... 81cf69c7c09
00000020  38:31:38:31:62:36:63:66:36:64:65:38:30:32:37:36  8181b6cf6de80276
00000030  35:61:61:39:33                                   5aa93

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0723, network: 0.0001s)
00000000  00:00:01:00:00:00:00:15:00:00:00:11:65:00:00:00  ............e...
00000010  01:00:00:00:00:00:00:00:00:00:00:00:00           .............

**Comment by terrafrost** [comment]
> My SSH2 logs turned out to be 105MB, but perhaps it's because I keep appending the log to the file.

So I misspoke. SSH2.php doesn't cap the size of the log file to 1MB - it caps the size of the data _being logged_ to 1MB. That's prior to converting it to hex and adding white space and what not to it to make it human readable. The log file will understandably be a good amount larger than the data being logged. 105MB is larger than I'd expect but I guess it's not an impossibility. Especially if you have a bunch of packets that are one byte long each since each of those would have the packet name, etc.

As for the packets you included in your post...  nothing looks suspicious there. The last outbound NET_SSH2_MSG_CHANNEL_DATA packet is an SSH_FXP_CLOSE. The last inbound packet is SSH_FXP_STATUS.

Between the line number you gave and the packet log it's not immediately obvious to me what's causing that error. After a file has been closed I'd expect it to be the client that sends out the next SFTP packet - not the server. So why is phpseclib expecting another packet?

Maybe you could do this:

```
#
#-----[ OPEN ]------------------------------------------
#
Net/SFTP.php
#
#-----[ FIND ]------------------------------------------
#
user_error('Invalid SFTP packet size');
#
#-----[ REPLACE WITH ]----------------------------------
#
throw new \RuntimeException('Invalid SFTP packet size');
```
That should give a stack trace.

**Comment by dingman** [comment]
I'm getting this error as well, the latest version has the Exception already being thrown.

```10	1.5103	7818504	phpseclib3\Net\SFTP->nlist( )	...\Remote.php:28
11	1.5103	7818504	phpseclib3\Net\SFTP->nlist_helper( )	...\SFTP.php:892
12	1.5103	7818504	phpseclib3\Net\SFTP->readlist( )	...\SFTP.php:906
13	1.5608	7818568	phpseclib3\Net\SFTP->get_sftp_packet( )	...\SFTP.php:983
```

**Comment by terrafrost** [comment]
It is not clear to me what version of phpseclib you're running. I mean, you're running 3.0, but which version of 3.0, I can't tell.

Line 892 doesn't call `nlist_helper` in any tagged version. In [3.0.6](https://github.com/phpseclib/phpseclib/blob/3.0.6/phpseclib/Net/SFTP.php#L892) it's `if (!($this->bitmap & SSH2::MASK_LOGIN)) {` and in [every other version](https://github.com/phpseclib/phpseclib/blob/3.0.5/phpseclib/Net/SFTP.php#L892) it's `return false;`

Anyway, `get_sftp_packet` is called twice in `readlist`:

https://github.com/phpseclib/phpseclib/blob/3.0.5/phpseclib/Net/SFTP.php#L905

https://github.com/phpseclib/phpseclib/blob/3.0.5/phpseclib/Net/SFTP.php#L933

Which call is causing the issue?

Also, some context would be helpful.

The other person who was having the issue was "_downloading 100s of files using $sftp->get() and running it in a loop to download each file one by one from a remote directory_". Are you doing something similar?

The SSH logs might be helpful. You can get them by doing `define('NET_SSH2_LOGGING', 2)` and then by doing `echo $sftp->getLog()` after the fact.

If the problem is consistently reproducible on a particular system could I have access to that system so that I might be able to consistently reproduce the issue myself? If you were up for this you could email credentials to terrafrost@php.net.

**Comment by dingman** [comment]
I'm running latest release, 3.0.6

I must have provided the wrong debug then, my apologies.

The line that is getting called is 2980 in SFTP.php

```        
        if ($tempLength > 256 * 1024) {
            throw new \RuntimeException('Invalid Size');
        }
```

I am also doing something similar, yes. Maybe not 100s, but at least 10s of files through a loop.

I'm using nlist to get the remote list, then running logic to decide which files to download.

It happens immediately, on every attempt to run the logic to download any file.

It's on my local machine, so unfortunately no access.

I implemented those two lines of code, however it's not rendering anything because of the fatal error before getting to that line of code.

```

( ! ) Fatal error: Uncaught RuntimeException: Invalid Size in C:\Users\Jonathan\Local Sites\remotefiletesting\app\public\wp-content\plugins\test\vendor\phpseclib\phpseclib\phpseclib\Net\SFTP.php on line 2980
--

( ! ) RuntimeException: Invalid Size in C:\Users\Jonathan\Local Sites\remotefiletesting\app\public\wp-content\plugins\test\vendor\phpseclib\phpseclib\phpseclib\Net\SFTP.php on line 2980
--
```

```
	public function download( $remoteFilename ) {
		$remoteBlob = $this->sftp->get( $remoteFilename );
		echo $this->sftp->getLog();
		dd( "stop" );
		$this->tmpFile         = new File( $remoteBlob, $remoteFilename );
		$this->tmpFile->delete = FALSE;

		return $this;
	}
```

**Comment by terrafrost** [comment]
> I implemented those two lines of code, however it's not rendering anything because of the fatal error before getting to that line of code

Try catching that error and then outputting the logs. eg.

```php
try {
    $files = $sftp->nlist();
} catch (\Exception $e) {
    echo $sftp->getLog();
}
```

**Comment by ricohumme** [comment]
Hi,

I'm currently experiencing the same issues on version 2.0.31
Here I also get the full listing of a directory so I can apply logic as to which files I need to still upload to the remote server.
The error thrown here is `User Notice: Invalid SFTP packet size`
As you requested the logs from others I thought I'd provide you with them directly.
Due to possible sensitive data, I have sent the logs to your email address listed on your profile.

Thank you in advance!

Kind regards,
Rico

**Comment by ricohumme** [comment]
By commenting the error which is thrown I can proceed with the request.
There are about 7000+ files in the listing on the remote server.

Is it possible that the size of the listing is exceeds the SFTP_MAX_MSG_LENGTH?
```
        extract(unpack('Nlength', $this->_string_shift($this->packet_buffer, 4)));
        $tempLength = $length;
        $tempLength-= strlen($this->packet_buffer);


        // 256 * 1024 is what SFTP_MAX_MSG_LENGTH is set to in OpenSSH's sftp-common.h
        if ($tempLength > 256 * 1024) {
//            user_error('Invalid SFTP packet size');
//            return false;
        }
```
The threshold here is 262144 while my value of `tempLength` exceeds well over 1 million.
However, with the above lines commented out, it still works, so why is the error thrown in this case?

**Comment by terrafrost** [comment]
Sorry for the delay. That's some nice sleuthing on your part!

So for that I'm basically just copying the behavior of PuTTY and OpenSSH. No doubt the idea is that if a decryption fails that the size will be completely non-sensical.

#1371 is why that line was changed.

Can you download the files from that server with, say, WinSCP? (PuTTY powers WinSCP)

I note from your logs that you're using https://www.wftpserver.com/ . Imma try to install that and verify that it works on that and then I can add that as something special that phpseclib does just for that server. I can also verify OpenSSH's / PuTTY's behavior.

**Comment by terrafrost** [comment]
I'm able to reproduce the issue!

**Comment by ricohumme** [comment]
That's nice. If there is a fix I can try, please let me know

**Comment by terrafrost** [comment]
I haven't forgotten about this. I've just been super busy these past few days. Hopefully I'll have some time this weekend!

**Comment by terrafrost** [comment]
https://github.com/phpseclib/phpseclib/commit/0673370edace281f1345d7411e0ec65406835530 should fix this.

Thanks!

**Comment by Clevero** [comment]
Seems like I'm also facing this issue with 3.0.9.
Are there any ETA for a new release with this fix?

```
Fatal error: Uncaught RuntimeException: Invalid Size in /var/www/html/vendor/phpseclib/phpseclib/phpseclib/Net/SFTP.php on line 2963

RuntimeException: Invalid Size in /var/www/html/vendor/phpseclib/phpseclib/phpseclib/Net/SFTP.php on line 2963
```

**Comment by terrafrost** [comment]
It looks like this fix was implemented the day after 3.0.9 was released.

I'm going to be out of town this weekend but maybe next weekend I can do a new release.

Thanks!

**Comment by Clevero** [comment]
Thank you :)
